### PR TITLE
Refactor error code to support warnings

### PR DESCRIPTION
**Pull request by omnivion**
### Information
- JIRA story QPPCT-1019.

### Checklist 
- [x] All JUnit tests pass (`mvn clean verify`).
- [x] New unit tests written to cover new functionality.
- [x] Added and updated JavaDocs for non-test classes and methods.
- [x] No local design debt. Do you feel that something is "ugly" after your changes?
- [x] Updated documentation (`README.md`, etc.) depending if the changes require it.

### CODE REVIEW

File: converter/src/main/java/gov/cms/qpp/conversion/ConversionReport.java [hunk 1]
```diff
@@ -52,10 +57,13 @@ private AllErrors constructErrorHierarchy(final String inputIdentifier, final Li
 	}
 
 	/**
-	 * Constructs an {@link Error} for the given {@code inputIdentifier} from the passed in validation errors.
+	 * Constructs an {@link Error} for the given {@code inputIdentifier} from the
+	 * passed in validation errors.
 	 *
-	 * @param inputIdentifier An identifier for a source of QRDA3 XML.
-	 * @param details A list of validation errors.
+	 * @param inputIdentifier
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/ConversionReport.java]: Pourquoi?

File: converter/src/main/java/gov/cms/qpp/conversion/ConversionReport.java [hunk 2]
```diff
@@ -139,6 +149,14 @@ public Source getValidationErrorsSource() {
 		}
 	}
 
+	public List<Detail> getWarnings() {
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/ConversionReport.java]: <indignation> WHAT!?! </indignation> no javadocs?

File: converter/src/main/java/gov/cms/qpp/conversion/Converter.java [hunk 3]
```diff
@@ -158,7 +163,7 @@ protected JsonOutputEncoder getEncoder() {
 	 * @return the conversion report
 	 */
 	public ConversionReport getReport() {
-		return new ConversionReport(source, details, decoded, encoded);
+		return new ConversionReport(source, errors, warnings, decoded, encoded);
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/Converter.java]: maybe a builder here?

File: converter/src/main/java/gov/cms/qpp/conversion/validate/CommonNumeratorDenominatorValidator.java [hunk 4]
```diff
@@ -21,9 +21,9 @@
 	 * @param node Node parsed xml fragment under consideration
 	 */
 	@Override
-	protected void internalValidateSingleNode(Node node) {
-		check(node).childExact(format(ErrorCode.NUMERATOR_DENOMINATOR_CHILD_EXACT), 1, TemplateId.PI_AGGREGATE_COUNT);
-		if (getDetails().isEmpty()) {
+	protected void performValidation(Node node) {
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/CommonNumeratorDenominatorValidator.java]: 👍

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/ValidationResult.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
+package gov.cms.qpp.conversion.model.error;
+
+import java.util.List;
+
+public class ValidationResult {
```
Comment (sfradkin) [inline converter/src/main/java/gov/cms/qpp/conversion/model/error/ValidationResult.java]: I'm sure @clydet will ask for some class level javadoc...

File: converter/src/main/java/gov/cms/qpp/conversion/validate/NodeValidator.java [hunk 6]
```diff
@@ -1,72 +1,70 @@
...
+import gov.cms.qpp.conversion.model.error.ValidationResult;
 
 /**
  * The parent class that all validators must inherit from.
  */
 public abstract class NodeValidator {
 
 	private static final Logger DEV_LOG = LoggerFactory.getLogger(NodeValidator.class);
-	private Set<Detail> details = new LinkedHashSet<>();
+	private final List<Detail> errors = new ArrayList<>();
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/NodeValidator.java]: why are we going with lists now?

File: converter/src/main/java/gov/cms/qpp/conversion/model/error/ValidationResult.java (hunk 5 above)
Comment (clydet) [reply converter/src/main/java/gov/cms/qpp/conversion/model/error/ValidationResult.java]: soooo how 'bout a class level javadoc comment?

**Comment by sfradkin** [comment]
I didn't see any unit tests specifically verifying the behavior of the new warnings List.

File: converter/src/main/java/gov/cms/qpp/conversion/validate/NodeValidator.java [hunk 7]
```diff
@@ -1,72 +1,70 @@
...
+	/**
+	 * Used by child classes to add a {@link Detail}.
+	 *
+	 * @param newError The error to add to the list.
+	 */
+	protected final void addError(Detail error) {
+		errors.add(error);
+	}
+
+	public boolean containsError(Detail detail) {
```
Comment (clydet) [inline converter/src/main/java/gov/cms/qpp/conversion/validate/NodeValidator.java]: javadocs

**Comment by clydet** [comment]
lgtm

**Pull request merged**